### PR TITLE
Handle layer's negative top or left

### PR DESCRIPTION
**Pull request by namse**
closes https://github.com/chinedufn/psd/issues/45

Other little code changes by rust-analyzer and rustfmt automatically.

### CODE REVIEW

**Comment by namse** [comment]
Hmm. the problem is very complex here.

`rgba_idx` could be negative if the layer's top and left is negative. It's not considered in code.

**Comment by namse** [comment]
I put the change that `fn rgba_idx(...)` returns `Option<usize>` at commit [e4a008b](https://github.com/chinedufn/psd/pull/46/commits/e4a008b5fdba07a58f972480b34cba41306cded1).

File: src/render.rs [hunk 1]
```diff
@@ -74,10 +74,10 @@ impl<'a> Renderer<'a> {
         for (idx, layer) in self.layers_to_flatten_top_down.iter().enumerate() {
             // If this pixel is out of bounds of this layer we return the pixel below it.
             // If there is no pixel below it we return a transparent pixel
-            if pixel_left < layer.layer_properties.layer_left as usize
-                || pixel_left > layer.layer_properties.layer_right as usize
-                || pixel_top < layer.layer_properties.layer_top as usize
-                || pixel_top > layer.layer_properties.layer_bottom as usize
```
Comment (namse) [inline src/render.rs:80]: layer_left, right, top, bottom could be negative, so pixel_left and top should be casted from usize to i32, not casting i32 to usize.

File: src/psd_channel.rs [hunk 2]
```diff
@@ -203,7 +209,7 @@ fn rle_decompress(bytes: &[u8]) -> Vec<u8> {
         } else {
             let repeat = 1 - header;
             let byte = cursor.read_1()[0];
-            for _ in 0..repeat as usize {
+            for _ in 0..repeat {
```
Comment (namse) [inline src/psd_channel.rs:212]: I removed this kind of meaningless `as usize` codes.

File: src/sections/image_data_section.rs [hunk 3]
```diff
@@ -32,16 +32,16 @@ pub enum ImageDataSectionError {
 #[derive(Debug)]
 pub struct ImageDataSection {
     /// The compression method for the image.
-    pub(in crate) compression: PsdChannelCompression,
+    pub(crate) compression: PsdChannelCompression,
```
Comment (namse) [inline src/sections/image_data_section.rs:35]: This changed by rustfmt and rust-analzyer automatically.

File: src/sections/image_data_section.rs [hunk 4]
```diff
@@ -161,7 +161,7 @@ impl ImageDataSection {
                 let (red_start, red_end) =
                     (channel_data_start, channel_data_start + red_byte_count);
 
-                let red = bytes[red_start as usize..red_end as usize].into();
+                let red = bytes[red_start..red_end].into();
```
Comment (namse) [inline src/sections/image_data_section.rs:164]: these guys are already usize.

File: src/sections/layer_and_mask_information_section/layer.rs [hunk 5]
```diff
@@ -429,13 +429,20 @@ impl IntoRgba for PsdLayer {
...
-    fn rgba_idx(&self, idx: usize) -> usize {
-        let left_in_layer = idx % self.width() as usize;
-        let left_in_psd = self.layer_properties.layer_left as usize + left_in_layer;
+    fn rgba_idx(&self, idx: usize) -> Option<usize> {
+        let left_in_layer = idx as u16 % self.width();
+        let left_in_psd = self.layer_properties.layer_left + left_in_layer as i32;
 
-        let top_in_psd = idx / self.width() as usize + self.layer_properties.layer_top as usize;
+        let top_in_layer = idx as u16 / self.width();
+        let top_in_psd = self.layer_properties.layer_top + top_in_layer as i32;
```
Comment (namse) [inline src/sections/layer_and_mask_information_section/layer.rs]: I casted unsigned integer to signed integer,
previously this casted signed integer to unsigned integer, which can occur overflow.

**Comment by chinedufn** [comment]
Thanks for working on this! Will review this over the weekend.

**Comment by namse** [comment]
I see, I have a psd file that makes error with previous code. Give me some time to put test. :)

**Comment by kevzettler** [comment]
@namse I created a PR to add tests to your branch https://github.com/namse/psd/pull/1

@chinedufn this PR seems to fix few couple issues #45 #26 #43  and would be good to get in for a release

**Review by chinedufn — APPROVED** [review]
Thanks! Just need to land @kevzettler 's tests then I can merge this.

**Comment by kevzettler** [comment]
@namse @chinedufn I have created a new PR at #52 which contains all the changes from this PR with the test cases. We can close this in favor of #52

**Comment by chinedufn** [comment]
Added a commit with both of you as co authors https://github.com/chinedufn/psd/commit/a57f32fc5d300e3a5398aa313b74cb4461907f69

**Pull request closed without merging**